pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/SourceSinkUtils.java

```diff
@@ -83,21 +83,37 @@ private static Map<String, String> validateSourceOptions(
         return caseInsensitiveParams;
     }
 
-    public static boolean belongsTo(TopicRange topicRange, int numParallelSubtasks, int index) {
-        String topic = topicRange.getTopic();
-        if (topic.contains(PulsarOptions.PARTITION_SUFFIX)) {
-            int pos = topic.lastIndexOf(PulsarOptions.PARTITION_SUFFIX);
-            String topicPrefix = topic.substring(0, pos);
-            String topicPartitionIndex =
-                    topic.substring(pos + PulsarOptions.PARTITION_SUFFIX.length());
-            if (topicPartitionIndex.matches("0|[1-9]\\d*")) {
-                int startIndex =
-                        (topicPrefix.hashCode() * 31 & Integer.MAX_VALUE) % numParallelSubtasks;
-                return (startIndex + Integer.valueOf(topicPartitionIndex)) % numParallelSubtasks
-                        == index;
+    /**
+     * Decide if a topic should be consumed by the subTask with given index and range. If the range
+     * is a full range, then the topic belongs to only one subTask If the range is not a full range,
+     * then the topic belongs to multiple subTasks with each of them be responsible for part of the
+     * range. This happens only when `enable-key-hash-range` is set.
```

Comment:
Assume there are 5 subtasks,  and key hash enabled. Is a topic guaranteed to be split into 5 key hash range and each of the subtask has one of them ?

Reply:
You need to view it the other way around:

Each SubTask register to the pulsar that it will handle part of the topic.